src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,296 @@
+	EntryAdapter,
+	IGrpcClientAsync,
+	promisifyClient,
+	PusherClient,
+	PushRequest,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	status,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { errors } from '@balena/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = errors;
+
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: IGrpcClientAsync;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = promisifyClient(
+			new PusherClient(
+				`${LOKI_HOST}:${LOKI_PORT}`,
+				createInsecureCredentials(),
+			),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const oneHourAgo = new Date(Date.now() - 10000 * 60);
+
+		const queryRequest = new QueryRequest();
+		queryRequest.setSelector(this.getDeviceQuery(ctx));
+		queryRequest.setLimit(Number.isFinite(count) ? count : 1000);
+		queryRequest.setStart(createTimestampFromDate(oneHourAgo));
+		queryRequest.setEnd(createTimestampFromDate());
+		queryRequest.setDirection(Direction.BACKWARD);
+
+		const streams: StreamAdapter[] = [];
+		const call = this.querier.query(
+			queryRequest,
+			createOrgIdMetadata(String(ctx.belongs_to__application)),
+		);
+		const responseStreams: StreamAdapter[] = await new Promise(
+			(resolve, reject) => {
+				call.on('data', (queryResponse: QueryResponse) => {
+					streams.push(...queryResponse.getStreamsList());
+				});
+				call.on('error', (error: Error & { details: string }) => {
+					captureException(error);
```

Comment:
Can you add some blurb to the captureException so it's a bit easier to understand where it comes from when looking in sentry

Reply:
Updated